File: server/src/utils/dedup.ts
import * as crypto from "crypto";

const SIMHASH_BITS = 64;
const SIMHASH_THRESHOLD_85_PERCENT = 10; // 64-bit, 10 bits different = 84.4% similarity

const SHORT_TEXT_BOUNDARY = 100;

function normalize(text: string): string {
    return text
        .replace(/[\s，。！？、；：""''（）《》【】…—\u3000]/g, "")
        .replace(/\s+/g, "")
        .toLowerCase();
}

function md5(text: string): string {
    return crypto.createHash("md5").update(text, "utf8").digest("hex");
}

function tokenize(text: string): string[] {
    const cleaned = normalize(text);
    if (cleaned.length === 0) return [];
    const tokens: string[] = [];
    for (let i = 0; i < cleaned.length - 1; i++) {
        tokens.push(cleaned.slice(i, i + 2));
    }
    if (tokens.length === 0 && cleaned.length > 0) {
        tokens.push(cleaned);
    }
    return tokens;
}

function stringHash(str: string): number {
    let hash = 5381;
    for (let i = 0; i < str.length; i++) {
        hash = (hash * 33) ^ str.charCodeAt(i);
    }
    return Math.abs(hash);
}

function intToBinaryString(num: number, bits: number): string {
    return (num >>> 0).toString(2).padStart(bits, "0").slice(-bits);
}

function computeSimHash(text: string): string {
    if (!text) return "";
    const tokens = tokenize(text);
    const v: number[] = new Array(SIMHASH_BITS).fill(0);

    for (const token of tokens) {
        const hash = stringHash(token);
        const binary = intToBinaryString(hash, SIMHASH_BITS);
        for (let i = 0; i < SIMHASH_BITS; i++) {
            if (binary[i] === "1") v[i] += 1;
            else v[i] -= 1;
        }
    }

    let fingerprint = "";
    for (let i = 0; i < SIMHASH_BITS; i++) {
        fingerprint += v[i] >= 0 ? "1" : "0";
    }
    return fingerprint;
}

function hammingDistance(a: string, b: string): number {
    if (a.length !== b.length) return SIMHASH_BITS;
    let dist = 0;
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) dist++;
    }
    return dist;
}

function similarityPercent(a: string, b: string): number {
    if (!a || !b) return 0;
    const dist = hammingDistance(a, b);
    return ((SIMHASH_BITS - dist) / SIMHASH_BITS) * 100;
}

export interface DedupInput {
    content: string;
}

export interface DedupFingerprint {
    contentMd5: string;
    simHash: string;
    normalized: string;
}

export interface DedupContext {
    knownMd5s: Set<string>;
    knownSimHashes: string[];
    threshold: number; // hamming distance threshold for 85% similarity
}

function fingerprint(content: string): DedupFingerprint {
    const normalized = normalize(content);
    return {
        contentMd5: md5(normalized),
        simHash: computeSimHash(content),
        normalized,
    };
}

function isDuplicate(content: string, ctx: DedupContext, contentLen?: number): { isDup: boolean; reason?: "md5" | "simhash" } {
    const len = contentLen ?? content.length;
    const fp = fingerprint(content);

    // 1. MD5 精确匹配（短文本唯一手段，长文本也作为快速路径）
    if (ctx.knownMd5s.has(fp.contentMd5)) {
        return { isDup: true, reason: "md5" };
    }

    // 2. 长文本才做 SimHash 相似度检查
    if (len >= SHORT_TEXT_BOUNDARY) {
        for (const existingHash of ctx.knownSimHashes) {
            if (!existingHash) continue;
            const dist = hammingDistance(fp.simHash, existingHash);
            if (dist <= ctx.threshold) {
                return { isDup: true, reason: "simhash" };
            }
        }
    }

    return { isDup: false };
}

export const dedup = {
    SIMHASH_BITS,
    SHORT_TEXT_BOUNDARY,
    DEFAULT_SIMHASH_THRESHOLD: SIMHASH_THRESHOLD_85_PERCENT,

    normalize,
    md5,
    computeSimHash,
    hammingDistance,
    similarityPercent,
    fingerprint,
    isDuplicate,
};

export default dedup;

File: server/app/utils/dedup.py
"""分级去重引擎（MD5 + SimHash）

对应 TypeScript 版 server/src/utils/dedup.ts
短文本（<100字）用 MD5 精确匹配；长文本额外做 SimHash 相似度检查。
"""
import hashlib
import re

SIMHASH_BITS = 64
SIMHASH_THRESHOLD_85_PERCENT = 10  # 64-bit, 10 bits different = 84.4% similarity
SHORT_TEXT_BOUNDARY = 100

_PUNCT_RE = re.compile(r"[\s，。！？、；：""''（）《》【】…—\u3000]")


def normalize(text: str) -> str:
    t = _PUNCT_RE.sub("", text)
    t = re.sub(r"\s+", "", t)
    return t.lower()


def md5(text: str) -> str:
    return hashlib.md5(text.encode("utf-8")).hexdigest()


def tokenize(text: str):
    cleaned = normalize(text)
    if not cleaned:
        return []
    tokens = []
    for i in range(len(cleaned) - 1):
        tokens.append(cleaned[i:i + 2])
    if not tokens and len(cleaned) > 0:
        tokens.append(cleaned)
    return tokens


def string_hash(s: str) -> int:
    """DJB hash 变体（与 TS 版一致：hash*33 ^ charCode，取绝对值）"""
    h = 5381
    for ch in s:
        h = (h * 33) ^ ord(ch)
    return abs(h)


def int_to_binary_string(num: int, bits: int) -> str:
    # 模拟 JS (num >>> 0).toString(2) 的 32 位无符号行为
    num = num & 0xFFFFFFFF
    s = bin(num)[2:]
    return s.rjust(bits, "0")[-bits:]


def compute_sim_hash(text: str) -> str:
    if not text:
        return ""
    tokens = tokenize(text)
    v = [0] * SIMHASH_BITS
    for token in tokens:
        h = string_hash(token)
        binary = int_to_binary_string(h, SIMHASH_BITS)
        for i in range(SIMHASH_BITS):
            if binary[i] == "1":
                v[i] += 1
            else:
                v[i] -= 1
    fp = "".join("1" if x >= 0 else "0" for x in v)
    return fp


def hamming_distance(a: str, b: str) -> int:
    if len(a) != len(b):
        return SIMHASH_BITS
    return sum(1 for x, y in zip(a, b) if x != y)


def similarity_percent(a: str, b: str) -> float:
    if not a or not b:
        return 0
    dist = hamming_distance(a, b)
    return ((SIMHASH_BITS - dist) / SIMHASH_BITS) * 100


def fingerprint(content: str):
    """返回 { contentMd5, simHash, normalized }"""
    normalized = normalize(content)
    return {
        "contentMd5": md5(normalized),
        "simHash": compute_sim_hash(content),
        "normalized": normalized,
    }


def is_duplicate(content: str, known_md5s: set, known_sim_hashes: list, threshold: int = SIMHASH_THRESHOLD_85_PERCENT, content_len: int | None = None):
    """判重，返回 (isDup, reason|None)"""
    length = content_len if content_len is not None else len(content)
    fp = fingerprint(content)

    if fp["contentMd5"] in known_md5s:
        return True, "md5"

    if length >= SHORT_TEXT_BOUNDARY:
        for existing in known_sim_hashes:
            if not existing:
                continue
            if hamming_distance(fp["simHash"], existing) <= threshold:
                return True, "simhash"

    return False, None


dedup = {
    "SIMHASH_BITS": SIMHASH_BITS,
    "SHORT_TEXT_BOUNDARY": SHORT_TEXT_BOUNDARY,
    "DEFAULT_SIMHASH_THRESHOLD": SIMHASH_THRESHOLD_85_PERCENT,
    "normalize": normalize,
    "md5": md5,
    "computeSimHash": compute_sim_hash,
    "hammingDistance": hamming_distance,
    "similarityPercent": similarity_percent,
    "fingerprint": fingerprint,
    "isDuplicate": is_duplicate,
}
